firebase-perf/src/main/java/com/google/firebase/perf/util/Timer.java

```diff
@@ -14,92 +14,105 @@
 
 package com.google.firebase.perf.util;
 
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
 import android.os.Parcel;
 import android.os.Parcelable;
+import android.os.SystemClock;
 import androidx.annotation.NonNull;
 import com.google.android.gms.common.util.VisibleForTesting;
-import java.util.concurrent.TimeUnit;
 
-/** A Timer class provides both wall-clock (epoch) time and high resolution time (nano time). */
+/**
+ * A Timer class provides both wall-clock (epoch) time and monotonic time (elapsedRealtime).
+ * Milliseconds is the time unit used in private fields and computations, because that's the time
+ * unit most widely available in Android APIs. Public methods only return time in microseconds,
+ * because Fireperf proto requires microseconds.
+ */
 public class Timer implements Parcelable {
 
-  /** Wall-clock time or epoch time in microseconds, */
-  private long timeInMicros;
   /**
-   * High resolution time in nanoseconds. High resolution time should only be used to calculate
-   * duration or latency. It is not wall-clock time.
+   * Wall-clock time or epoch time in microseconds. Do NOT use for duration because wall-clock is
+   * not guaranteed to be monotonic: it can be set by the user or the phone network, thus it may
+   * jump forwards or backwards unpredictably. {@see SystemClock}
    */
-  private long highResTime;
-
+  private long wallClock;
   /**
-   * Construct Timer object using System clock. Make it package visible to be only accessible from
-   * com.google.firebase.perf.util.Clock.
+   * Monotonic time measured in the {@link SystemClock#elapsedRealtime()} timebase. Only used to
+   * compute duration between 2 timestamps in the same timebase. It is NOT wall-clock time.
    */
-  public Timer() {
-    timeInMicros = TimeUnit.MILLISECONDS.toMicros(System.currentTimeMillis());
-    highResTime = System.nanoTime();
-  }
+  private long elapsedRealtime;
 
   /**
-   * Construct a Timer object with input wall-clock time, assume high resolution time is same as
-   * wall-clock time.
+   * Returns a new Timer object as if it was stamped at the given elapsedRealtime. Uses current
+   * wall-clock as a reference to extrapolate the wall-clock at the given elapsedRealtime.
    *
-   * @param time wall-clock time in microseconds
+   * @param elapsedRealtime timestamp in the {@link SystemClock#elapsedRealtime()} timebase
    */
-  @VisibleForTesting
-  public Timer(long time) {
-    this.timeInMicros = time;
-    highResTime = TimeUnit.MICROSECONDS.toNanos(time);
+  public static Timer ofElapsedRealtime(long elapsedRealtime) {
+    Timer now = new Timer();
```

Comment:
I wanted to encapsulate the responsibility of what APIs are used to capture "now" for both wall-clock and monotonic time to the Timer constructor (currently we use `System.currentTimeMillis()` and `SystemClock.elapsedRealtime()` but what if we change it in the future when future APIs come around? then we have to change everywhere these APIs are called, as opposed to only changing 1 constructor). 

As for memory, it's just a local variable in a short method that will be marked for garbaged collection as soon as this method finishes, so I don't personally see a concern.

Reply:
This seems like it should be a builder pattern or calling a separate protected method to re-assign these variables. It seems unusual to need to end up creating 2 instances just to reassign variables, especially since we are writing the class.

Reply:
Decided to change it to use Android APIs still, because it's impossible to completely encapsulate the timebase. `ofElapsedRealtime(...)` method itself requires knowledge of the `elapsedRealtime` timebase anyways, so might as well keep it coupled.

Reply:
I just spoke with Julio and he thinks there's also a code smell with having the two timers. His suggestions were to either use a separate constructor (and not use the factory) like Timer(long elapsedRealtime) and have that perform the calculation and potentially have Timer() call Timer(long elapsedRealtime).

Alternatively, having the system calls in the calculations to avoid the double objects being constructed.

Reply:
Updated and added 2 tests for `ofElapsedRealtime(long)`, @raymondlam PTAL.